flocker/control/test/test_protocol.py

```diff
@@ -0,0 +1,212 @@
+        as_bytes = argument.toString(TEST_DEPLOYMENT)
+        deserialized = argument.fromString(as_bytes)
+        self.assertEqual([bytes, TEST_DEPLOYMENT],
+                         [type(as_bytes), deserialized])
+
+
+class ControlServiceLocatorTests(SynchronousTestCase):
+    """
+    Tests for ``ControlServiceLocator``.
+    """
+    def setUp(self):
+        self.cluster_state = ClusterStateService()
+        self.cluster_state.startService()
+        self.addCleanup(self.cluster_state.stopService)
+        self.client = LoopbackAMPClient(ControlServiceLocator(
+            self.cluster_state))
+
+    def test_version(self):
+        """
+        ``VersionCommand`` to the control service returns the current internal
+        protocol version.
+        """
+        self.assertEqual(
+            self.successResultOf(self.client.callRemote(VersionCommand)),
+            {"major": 1})
+
+    def test_nodestate(self):
+        """
+        ``NodeStateCommand`` updates the node state.
+        """
+        self.successResultOf(
+            self.client.callRemote(NodeStateCommand, hostname=u"example1",
+                                   node_state=NODE_STATE))
+        self.assertEqual(self.cluster_state.as_deployment(),
+                         Deployment(
+                             nodes=frozenset([
+                                 Node(hostname=u'example1',
+                                      applications=frozenset([APP1, APP2]),
+                                      other_manifestations=frozenset(
+                                          [MANIFESTATION]))])))
+
+
+@implementer(IConvergenceAgent)
+@attributes([Attribute("is_connected", default_value=False),
+             Attribute("is_disconnected", default_value=False),
+             Attribute("desired", default_value=None),
+             Attribute("actual", default_value=None)])
+class FakeAgent(object):
+    """
+    Fake agent for testing.
+
+    Not a full verified fake since
+    https://clusterhq.atlassian.net/browse/FLOC-1255 may change this a
+    little.
+    """
+    def connected(self):
+        self.is_connected = True
+
+    def disconnected(self):
+        self.is_disconnected = True
+
+    def cluster_updated(self, configuration, cluster_state):
+        self.desired = configuration
+        self.actual = cluster_state
+
+
+class AgentClientTests(SynchronousTestCase):
+    """
+    Tests for ``AgentClient``.
+    """
+    def setUp(self):
+        self.agent = FakeAgent()
+        self.client = AgentClient(self.agent)
+        self.client.makeConnection(StringTransport())
+        # The server needs to send commands to the client, so it acts as
+        # an AMP client in that regard:
+        self.server = LoopbackAMPClient(self.client)
+
+    def test_connection_made(self):
+        """
+        Connection made events are passed on to the agent.
+        """
+        self.assertEqual(self.agent, FakeAgent(is_connected=True))
```

Comment:
This doesn't demonstrate that `is_connected` doesn't just start `True` and stay that way.